pkg/client/restclient/request.go

```diff
@@ -797,22 +798,31 @@ func (r *Request) request(fn func(*http.Request, *http.Response)) error {
 	// TODO: Change to a timeout based approach.
 	maxRetries := 10
 	retries := 0
+	var req *http.Request
+	var resp *http.Response
+	var u *url.URL
 	for {
-		url := r.URL().String()
-		req, err := http.NewRequest(r.verb, url, r.body)
-		if err != nil {
-			return err
-		}
-		req.Header = r.headers
+		err := r.urlProvider.WrapWithRetry(
```

Comment:
Ah this is probably better - moves the decision making on retries into the url provider.

Is there a design for how we intend retries to work? (whether or not we implement them in this PR).  We should probably look at e.g. how GRPC does it.

Or maybe we should just adopt GRPC and be done with it :-)